sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1645,3 +1646,30 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
       }
   }
 }
+
+/**
+ * Rewrite [[Filter]] plans that contain correlated [[ScalarSubquery]] expressions. When these
+ * correlated [[ScalarSubquery]] expressions are wrapped in a some Predicate expression, we rewrite
+ * them into [[PredicateSubquery]] expressions.
+ */
+object RewriteScalarSubqueriesInFilter extends Rule[LogicalPlan] {
```

Comment:
This rule is not necessary for this fix, right?

Also, we have a rule to turn outer join into semi join if possible, does that one not work in case?